src/std/os/signal-handler.ss

```diff
@@ -98,7 +92,93 @@ package: std/os
+      (lambda (self)
+        (let* ((handlers (make-vector SIGMAX #f))
+               (default-handlers (make-vector SIGMAX #f))
+               (kq (kqueue))
+               (mx (make-mutex 'signal-handler)))
+          (struct-instance-init! self handlers default-handlers kq mx)
+          (start-logger!)
+          (spawn/name 'signal-handler signal-handler-wait self))))
+
+    (def (signal-handler-wait sh)
+      (try
+        (defconst nevents 4)
+        (def events (make-kevents nevents))
+        (with ((signal-handler handlers default-handlers kq mx) sh)
+          (let wait-loop ()
+            (##wait-for-io! (fd-io-in kq) #t)
+            (let (n (kqueue-poll kq events nevents))
+              (when (##fxpositive? n)
+                (let event-loop ((i 0))
+                  (when (fx< i n)
+                    (if (##fxzero? (##fxand (kevent-flags events i)
+                                             EV_ERROR))
+                      (when (fx= (kevent-filter events i) EVFILT_SIGNAL)
+                        (let (event-handler
+                               (with-lock mx
+                                 (cut vector-ref handlers
+                                      (kevent-ident events i))))
+                          (when event-handler
+                            (signal-handler-dispatch event-handler))))
+                      (raise-os-error (kevent-data events i)
+                                      signal-handler-wait))
+                    (event-loop (fx1+ i))))))
+            (wait-loop)))
+        (catch (e)
+          (log-error "Error handling signals" e)
+          (raise e))))
+
+    (def (signal-handler-add! sh signo thunk)
+      (with ((signal-handler handlers default-handlers kq mx) sh)
+        (with-lock mx
+          (lambda ()
+            (let ((ev (make-kevents 1)))
+              (vector-set! handlers signo thunk)
+              (kqueue-kevent-add kq signo EVFILT_SIGNAL)
+              (ignore-signal! sh signo))))))
+
+    (def (signal-handler-remove! sh signo)
+      (with ((signal-handler handlers default-handlers kq mx) sh)
+        (with-lock mx
+          (lambda ()
+            (let ((ev (make-kevents 1)))
+              (restore-signal! sh signo)
+              (kqueue-kevent-del kq signo EVFILT_SIGNAL)
+              (vector-set! handlers signo #f))))))
+
+    (extern set-signal! signal-null-handler SIG_IGN)
+
+    (def (ignore-signal! sh signo)
+      (let* ((default-handlers (signal-handler-default-handlers sh))
+             (default-handler (set-signal! signo SIG_IGN)))
+        (unless (vector-ref default-handlers signo)
+          (vector-set! default-handlers signo default-handler))))
+
+    (def (restore-signal! sh signo)
+      (let* ((default-handlers (signal-handler-default-handlers sh))
+             (default-handler (vector-ref default-handlers signo)))
+        (when (vector-ref default-handlers signo)
+          (set-signal! signo default-handler))))
```

Comment:
shouldn't you also clear the default-handler here?

Reply:
There is still some cleaning to do. So it's not finished yet. I just pushed the merge to test it on the different BSDs. OpenBSD looks ok, but on NetBSD there are timeouts in the kqueue-server part. I am trying to figure out what's going on there.

The default-handlers were the trickier part in signal-handling. I tried to mimic the behaviour on Linux. A default-handler is put in the array the first time signal-handler-add! is called for a certain signal. Then it stays there forever and is used by `signal-handler-remove!` to restore the default behaviour. I would like to keep the default-handlers in the array, as otherwise the bookkeeping would be more complicated. But do you have a reason to clear them?

Reply:
that's fine, you don't have to clean it then; it was more of a consistency of interface thing.

Reply:
yeah, that's ok;  we can assume the application cedes control to the signal-handler library.

Reply:
So maybe we should leave the NetBSD/kqueue-server issues for a subsequent pr?
We can open a continuation issue.